# Executive Summary

this project analyzes one year of ecommerce data to identify sales concentration patterns, customer purchase repeat behavior, product purchasing dynamics, and operational opportunities that may support marketing allocation, and retention strategy decisions.

the analysis combines Python for data preparation, segmentation, and statistical validation with Tableau for visualization.

Key findings include:

- sales were heavily concentrated during Christmas period rather than across multiple major US holidays.
- revenue and repeat purchasing behavior were concentrated in a small number of cities, particularly California's cities.
- california's citiesdemonstrated substantially stronger repeat household concentration than other cities.
- repeat households generated significantly higher cumulative revenue than the one time purchase households, while average order value remained relatively stable across repeat segments.
- sales distribution across days of month remained relatively stable outside seasonal periods.
- the dataset does not include customer id, campaign attribution, profitability, customer lifecycle, or marketing spend data, limiting relational interpretation.


# Business Problem

The business wants to understaind the purchasing behaviour, sales concentration, customer behaviour in order to review:-
- bueinsess operations 
- aim for process and revenue optimizatoin

The project focuses on identifying whether:-
1- sales and customer buying behavior follow a specific patterns rather than a ordinary buying activity.

# Dataset Overview

The dataset contains  ecommerce sales records from 2019.

Main variables include:
- Order ID
- Product
- Quantity Ordered
- Price Each
- Order Date
- Purchase Address

The dataset contains approximately 186,305 rows.

Limitations:
- No customer id exists.
- No campaign, profitability, or acquisition source data exists.

# Data Cleaning & Preparation

The following preparation steps were performed before analysis:

- removed null records after verifying theier negligable effect on the analysis. 
- standardized column naming system were followed.
- converted quantity and price columns into numeric formats.
- created revenue metric:
    revenue = quantity_ordered × price_each
- extracted:
    - month
    - day
    - hour
    - city/state

- aggregated purchase addresses to approximate household behavior.
- created repeat household segmentation variables.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df1 = pd.read_csv('Sales/Sales_January_2019.csv')
df1['month']='Jan'
df2 = pd.read_csv('Sales/Sales_February_2019.csv')
df2['month']='Feb'
df3 = pd.read_csv('Sales/Sales_March_2019.csv')
df3['month']='Mar'
df4 = pd.read_csv('Sales/Sales_April_2019.csv')
df4['month']='Apr'
df5 = pd.read_csv('Sales/Sales_May_2019.csv')
df5['month']='May'
df6 = pd.read_csv('Sales/Sales_June_2019.csv')
df6['month']='Jun'
df7 = pd.read_csv('Sales/Sales_July_2019.csv')
df7['month']='Jul'
df8 = pd.read_csv('Sales/Sales_August_2019.csv')
df8['month']='Aug'
df9 = pd.read_csv('Sales/Sales_September_2019.csv')
df9['month']='Sep'
df10 = pd.read_csv('Sales/Sales_October_2019.csv')
df10['month']='Oct'
df11 = pd.read_csv('Sales/Sales_November_2019.csv')
df11['month']='Nov'
df12 = pd.read_csv('Sales/Sales_December_2019.csv')
df12['month']='Dec'
#adding new columns to each df

In [3]:
list=[df1, df2, df3, df4, df5, df6, df7, df8, df9, df10, df11, df12]

In [4]:
for i in list:
    print(i.shape)

(9723, 7)
(12036, 7)
(15226, 7)
(18383, 7)
(16635, 7)
(13622, 7)
(14371, 7)
(12011, 7)
(11686, 7)
(20379, 7)
(17661, 7)
(25117, 7)


In [5]:
frame = pd.concat(list, ignore_index=True)
frame

,Order ID,Product,Quantity Ordered,Price Each,Order Date,Purchase Address,month
0,141234,iPhone,1,700,01/22/19 21:25,"944 Walnut St, Boston, MA 02215",Jan
1,141235,Lightning Charging Cable,1,14.95,01/28/19 14:15,"185 Maple St, Portland, OR 97035",Jan
2,141236,Wired Headphones,2,11.99,01/17/19 13:33,"538 Adams St, San Francisco, CA 94016",Jan
3,141237,27in FHD Monitor,1,149.99,01/05/19 20:33,"738 10th St, Los Angeles, CA 90001",Jan
4,141238,Wired Headphones,1,11.99,01/25/19 11:59,"387 10th St, Austin, TX 73301",Jan
...,...,...,...,...,...,...,...
186845,319666,Lightning Charging Cable,1,14.95,12/11/19 20:58,"14 Madison St, San Francisco, CA 94016",Dec
186846,319667,AA Batteries (4-pack),2,3.84,12/01/19 12:01,"549 Willow St, Los Angeles, CA 90001",Dec
186847,319668,Vareebadd Phone,1,400,12/09/19 06:43,"273 Wilson St, Seattle, WA 98101",Dec
186848,319669,Wired Headphones,1,11.99,12/03/19 10:39,"778 River St, Dallas, TX 75001",Dec


In [6]:
frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186850 entries, 0 to 186849
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   Purchase Address  186305 non-null  object
 6   month             186850 non-null  object
dtypes: object(7)
memory usage: 10.0+ MB


In [7]:
frame.isnull().sum()

Order ID            545
Product             545
Quantity Ordered    545
Price Each          545
Order Date          545
Purchase Address    545
month                 0
dtype: int64

In [8]:
frame.duplicated().sum()

np.int64(1140)

In [9]:
frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186850 entries, 0 to 186849
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   Purchase Address  186305 non-null  object
 6   month             186850 non-null  object
dtypes: object(7)
memory usage: 10.0+ MB


In [10]:
frame.columns = [col.lower() for col in frame]
frame

,order id,product,quantity ordered,price each,order date,purchase address,month
0,141234,iPhone,1,700,01/22/19 21:25,"944 Walnut St, Boston, MA 02215",Jan
1,141235,Lightning Charging Cable,1,14.95,01/28/19 14:15,"185 Maple St, Portland, OR 97035",Jan
2,141236,Wired Headphones,2,11.99,01/17/19 13:33,"538 Adams St, San Francisco, CA 94016",Jan
3,141237,27in FHD Monitor,1,149.99,01/05/19 20:33,"738 10th St, Los Angeles, CA 90001",Jan
4,141238,Wired Headphones,1,11.99,01/25/19 11:59,"387 10th St, Austin, TX 73301",Jan
...,...,...,...,...,...,...,...
186845,319666,Lightning Charging Cable,1,14.95,12/11/19 20:58,"14 Madison St, San Francisco, CA 94016",Dec
186846,319667,AA Batteries (4-pack),2,3.84,12/01/19 12:01,"549 Willow St, Los Angeles, CA 90001",Dec
186847,319668,Vareebadd Phone,1,400,12/09/19 06:43,"273 Wilson St, Seattle, WA 98101",Dec
186848,319669,Wired Headphones,1,11.99,12/03/19 10:39,"778 River St, Dallas, TX 75001",Dec


In [11]:
frame.rename(columns={
    'order id': 'order_id',
    'quantity ordered': 'quantity_ordered',
    'price each':'price_each',
    'order date':'order_date',
    'purchase address':'purchase_address'}, inplace=True)
frame

,order_id,product,quantity_ordered,price_each,order_date,purchase_address,month
0,141234,iPhone,1,700,01/22/19 21:25,"944 Walnut St, Boston, MA 02215",Jan
1,141235,Lightning Charging Cable,1,14.95,01/28/19 14:15,"185 Maple St, Portland, OR 97035",Jan
2,141236,Wired Headphones,2,11.99,01/17/19 13:33,"538 Adams St, San Francisco, CA 94016",Jan
3,141237,27in FHD Monitor,1,149.99,01/05/19 20:33,"738 10th St, Los Angeles, CA 90001",Jan
4,141238,Wired Headphones,1,11.99,01/25/19 11:59,"387 10th St, Austin, TX 73301",Jan
...,...,...,...,...,...,...,...
186845,319666,Lightning Charging Cable,1,14.95,12/11/19 20:58,"14 Madison St, San Francisco, CA 94016",Dec
186846,319667,AA Batteries (4-pack),2,3.84,12/01/19 12:01,"549 Willow St, Los Angeles, CA 90001",Dec
186847,319668,Vareebadd Phone,1,400,12/09/19 06:43,"273 Wilson St, Seattle, WA 98101",Dec
186848,319669,Wired Headphones,1,11.99,12/03/19 10:39,"778 River St, Dallas, TX 75001",Dec


In [12]:
(frame.isnull().sum().sum()) / len(frame) * 100

np.float64(1.75006689858175)

In [13]:
frame = frame.dropna()
frame.isnull().sum()

order_id            0
product             0
quantity_ordered    0
price_each          0
order_date          0
purchase_address    0
month               0
dtype: int64

In [14]:
frame.info()

<class 'pandas.core.frame.DataFrame'>
Index: 186305 entries, 0 to 186849
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   order_id          186305 non-null  object
 1   product           186305 non-null  object
 2   quantity_ordered  186305 non-null  object
 3   price_each        186305 non-null  object
 4   order_date        186305 non-null  object
 5   purchase_address  186305 non-null  object
 6   month             186305 non-null  object
dtypes: object(7)
memory usage: 11.4+ MB


In [15]:
frame['quantity_ordered'].unique()

array(['1', '2', '3', '5', '4', '7', 'Quantity Ordered', '6', '9', '8'],
      dtype=object)

In [16]:
df = frame[frame['quantity_ordered'] != 'Quantity Ordered']

In [17]:
df['quantity_ordered'].unique()

array(['1', '2', '3', '5', '4', '7', '6', '9', '8'], dtype=object)

In [18]:
df = df.astype({'quantity_ordered': int, 'price_each': float})
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 185950 entries, 0 to 186849
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   order_id          185950 non-null  object 
 1   product           185950 non-null  object 
 2   quantity_ordered  185950 non-null  int64  
 3   price_each        185950 non-null  float64
 4   order_date        185950 non-null  object 
 5   purchase_address  185950 non-null  object 
 6   month             185950 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 11.3+ MB


In [19]:
df['revenue'] = df['price_each'] * df['quantity_ordered']
df

,order_id,product,quantity_ordered,price_each,order_date,purchase_address,month,revenue
0,141234,iPhone,1,700.00,01/22/19 21:25,"944 Walnut St, Boston, MA 02215",Jan,700.00
1,141235,Lightning Charging Cable,1,14.95,01/28/19 14:15,"185 Maple St, Portland, OR 97035",Jan,14.95
2,141236,Wired Headphones,2,11.99,01/17/19 13:33,"538 Adams St, San Francisco, CA 94016",Jan,23.98
3,141237,27in FHD Monitor,1,149.99,01/05/19 20:33,"738 10th St, Los Angeles, CA 90001",Jan,149.99
4,141238,Wired Headphones,1,11.99,01/25/19 11:59,"387 10th St, Austin, TX 73301",Jan,11.99
...,...,...,...,...,...,...,...,...
186845,319666,Lightning Charging Cable,1,14.95,12/11/19 20:58,"14 Madison St, San Francisco, CA 94016",Dec,14.95
186846,319667,AA Batteries (4-pack),2,3.84,12/01/19 12:01,"549 Willow St, Los Angeles, CA 90001",Dec,7.68
186847,319668,Vareebadd Phone,1,400.00,12/09/19 06:43,"273 Wilson St, Seattle, WA 98101",Dec,400.00
186848,319669,Wired Headphones,1,11.99,12/03/19 10:39,"778 River St, Dallas, TX 75001",Dec,11.99


# Hypothesis 1
sales increase during national holidays in the us specifically the most celebrated holidays, indicates the need for a dynamic inventory and campaign allocation strategies.
December 25 (Christmas Day)
November 28 (thanksgiving)
May 27, (memorial day)

In [20]:
monthly_sales = df.groupby('month')['revenue'].agg(sales_sum = 'sum').sort_values(by='sales_sum', ascending=False).reset_index()
monthly_sales

,month,sales_sum
0,Dec,4619297.12
1,Oct,3736884.05
2,Apr,3396059.11
3,Nov,3198909.23
4,May,3144584.80
5,Mar,2809063.30
6,Jul,2646899.69
7,Jun,2578293.30
8,Aug,2244412.31
9,Feb,2203481.24


In [21]:
def get_city(address):
    return address.split(',')[1].strip(' ')
def get_state(address):
    return address.split(',')[2].split(' ')[1]
df['city'] = df['purchase_address'].apply(lambda x: get_city(x) + ' ' + '('+ get_state(x)+')')
df

,order_id,product,quantity_ordered,price_each,order_date,purchase_address,month,revenue,city
0,141234,iPhone,1,700.00,01/22/19 21:25,"944 Walnut St, Boston, MA 02215",Jan,700.00,Boston (MA)
1,141235,Lightning Charging Cable,1,14.95,01/28/19 14:15,"185 Maple St, Portland, OR 97035",Jan,14.95,Portland (OR)
2,141236,Wired Headphones,2,11.99,01/17/19 13:33,"538 Adams St, San Francisco, CA 94016",Jan,23.98,San Francisco (CA)
3,141237,27in FHD Monitor,1,149.99,01/05/19 20:33,"738 10th St, Los Angeles, CA 90001",Jan,149.99,Los Angeles (CA)
4,141238,Wired Headphones,1,11.99,01/25/19 11:59,"387 10th St, Austin, TX 73301",Jan,11.99,Austin (TX)
...,...,...,...,...,...,...,...,...,...
186845,319666,Lightning Charging Cable,1,14.95,12/11/19 20:58,"14 Madison St, San Francisco, CA 94016",Dec,14.95,San Francisco (CA)
186846,319667,AA Batteries (4-pack),2,3.84,12/01/19 12:01,"549 Willow St, Los Angeles, CA 90001",Dec,7.68,Los Angeles (CA)
186847,319668,Vareebadd Phone,1,400.00,12/09/19 06:43,"273 Wilson St, Seattle, WA 98101",Dec,400.00,Seattle (WA)
186848,319669,Wired Headphones,1,11.99,12/03/19 10:39,"778 River St, Dallas, TX 75001",Dec,11.99,Dallas (TX)


In [22]:
def get_day(time):
    return time.split('/')[1].strip(' ')
df['day'] = df['order_date'].apply(lambda x: get_day(x))

def get_clock(time):
    return time.split('/')[2].split(' ')[1].split(':')[0]
df['hour'] = df['order_date'].apply(lambda x: get_clock(x))
df

,order_id,product,quantity_ordered,price_each,order_date,purchase_address,month,revenue,city,day,hour
0,141234,iPhone,1,700.00,01/22/19 21:25,"944 Walnut St, Boston, MA 02215",Jan,700.00,Boston (MA),22,21
1,141235,Lightning Charging Cable,1,14.95,01/28/19 14:15,"185 Maple St, Portland, OR 97035",Jan,14.95,Portland (OR),28,14
2,141236,Wired Headphones,2,11.99,01/17/19 13:33,"538 Adams St, San Francisco, CA 94016",Jan,23.98,San Francisco (CA),17,13
3,141237,27in FHD Monitor,1,149.99,01/05/19 20:33,"738 10th St, Los Angeles, CA 90001",Jan,149.99,Los Angeles (CA),05,20
4,141238,Wired Headphones,1,11.99,01/25/19 11:59,"387 10th St, Austin, TX 73301",Jan,11.99,Austin (TX),25,11
...,...,...,...,...,...,...,...,...,...,...,...
186845,319666,Lightning Charging Cable,1,14.95,12/11/19 20:58,"14 Madison St, San Francisco, CA 94016",Dec,14.95,San Francisco (CA),11,20
186846,319667,AA Batteries (4-pack),2,3.84,12/01/19 12:01,"549 Willow St, Los Angeles, CA 90001",Dec,7.68,Los Angeles (CA),01,12
186847,319668,Vareebadd Phone,1,400.00,12/09/19 06:43,"273 Wilson St, Seattle, WA 98101",Dec,400.00,Seattle (WA),09,06
186848,319669,Wired Headphones,1,11.99,12/03/19 10:39,"778 River St, Dallas, TX 75001",Dec,11.99,Dallas (TX),03,10


# Finding: 

- sales spikes were concentrated around christmas period while other holidays showed limited measurable impact, which could be an unexploited potential, if marketing campaign data existed to verify this. during the year there was spikes and drops in sale, which maybe attributed to many factors but not possible to verify without data.


- the business may benefit form scaling inventory, operational readiness and marketing campaigns specifically around the end of the year and the demand period rather than distributing resources evenly acroess the year.

# Limitation:
- no external compaign or promotional dataset exist to validate the drivers behind holidays current purshases behaviour and the spikes and drops across the year.

# Hypothesis 2

sales are disproportionately concentrated in a small number of cities, suggesting that geographically equal marketing allocation may lead to inefficient resource distribution.

In [23]:
sales_per_city = df.groupby('city').agg(revenue = ('revenue','sum')).sort_values('revenue', ascending=False).reset_index()
sales_per_city

,city,revenue
0,San Francisco (CA),8262203.91
1,Los Angeles (CA),5452570.80
2,New York City (NY),4664317.43
3,Boston (MA),3661642.01
4,Atlanta (GA),2795498.58
5,Dallas (TX),2767975.40
6,Seattle (WA),2747755.48
7,Portland (OR),1870732.34
8,Austin (TX),1819581.75
9,Portland (ME),449758.27


In [24]:
California_cities = ['San Francisco (CA)', 'Los Angeles (CA)']


df['group'] = df['city'].apply(lambda x: 'California_cities' if x in California_cities else 'Other Cities')

summary = df.groupby('group', as_index=False)['revenue'].sum()

summary['pct_of_total'] = summary['revenue'] / summary['revenue'].sum() * 100
summary

,group,revenue,pct_of_total
0,California_cities,13714774.71,39.762149
1,Other Cities,20777261.26,60.237851


In [39]:
summary.to_csv('summary_california_vs_other.csv', index=False)

## Findings

- sales and household concentration were heavily skewed toward california state cities contributing with around 40% of the total revenue of the company in the given year. 

- the findings suggest that geographic prioritization strategies may outperform evenly distributed marketing allocation approaches.
- the business need to invest in acquiring compaign data and collect more valuable data which would contribute to the understanding of human behaviour.
- verification process should be considered to point out why this behaviour is concentrating in california and how to duplicated it to other states.

## Limitation

- The dataset does not include regional marketing spend, customer acquisition source, or operational cost date to support the drivier behind the high revenue from the beforementioned cities.

# Hypothesis 3

household repeat purchasing behavior may indicate strong customer retention.

In [25]:
df['purchase_address'].duplicated().sum()

np.int64(45163)

In [26]:
household_df = df.groupby('purchase_address').agg({'order_id': 'nunique',
                                                   'revenue': 'sum',
                                                   'quantity_ordered': 'sum',
                                                   'city': 'first'}).reset_index()
household_df

,purchase_address,order_id,revenue,quantity_ordered,city
0,"1 11th St, Atlanta, GA 30301",1,11.95,1,Atlanta (GA)
1,"1 11th St, Los Angeles, CA 90001",1,1700.00,1,Los Angeles (CA)
2,"1 11th St, San Francisco, CA 94016",1,700.00,1,San Francisco (CA)
3,"1 12th St, Los Angeles, CA 90001",1,150.00,1,Los Angeles (CA)
4,"1 12th St, New York City, NY 10001",1,11.99,1,New York City (NY)
...,...,...,...,...,...
140782,"999 Wilson St, Atlanta, GA 30301",1,99.99,1,Atlanta (GA)
140783,"999 Wilson St, Los Angeles, CA 90001",1,999.99,1,Los Angeles (CA)
140784,"999 Wilson St, New York City, NY 10001",1,150.00,1,New York City (NY)
140785,"999 Wilson St, Portland, OR 97035",1,2.99,1,Portland (OR)


In [27]:
household_df.columns=['purchase_address', 'total_orders', 'total_revenue', 'total_quantity', 'city']
household_df

,purchase_address,total_orders,total_revenue,total_quantity,city
0,"1 11th St, Atlanta, GA 30301",1,11.95,1,Atlanta (GA)
1,"1 11th St, Los Angeles, CA 90001",1,1700.00,1,Los Angeles (CA)
2,"1 11th St, San Francisco, CA 94016",1,700.00,1,San Francisco (CA)
3,"1 12th St, Los Angeles, CA 90001",1,150.00,1,Los Angeles (CA)
4,"1 12th St, New York City, NY 10001",1,11.99,1,New York City (NY)
...,...,...,...,...,...
140782,"999 Wilson St, Atlanta, GA 30301",1,99.99,1,Atlanta (GA)
140783,"999 Wilson St, Los Angeles, CA 90001",1,999.99,1,Los Angeles (CA)
140784,"999 Wilson St, New York City, NY 10001",1,150.00,1,New York City (NY)
140785,"999 Wilson St, Portland, OR 97035",1,2.99,1,Portland (OR)


In [28]:
household_df['is_repeat'] = household_df['total_orders'] > 1
household_df['avg_order_value'] = household_df['total_revenue'] / household_df['total_orders']
household_df

,purchase_address,total_orders,total_revenue,total_quantity,city,is_repeat,avg_order_value
0,"1 11th St, Atlanta, GA 30301",1,11.95,1,Atlanta (GA),False,11.95
1,"1 11th St, Los Angeles, CA 90001",1,1700.00,1,Los Angeles (CA),False,1700.00
2,"1 11th St, San Francisco, CA 94016",1,700.00,1,San Francisco (CA),False,700.00
3,"1 12th St, Los Angeles, CA 90001",1,150.00,1,Los Angeles (CA),False,150.00
4,"1 12th St, New York City, NY 10001",1,11.99,1,New York City (NY),False,11.99
...,...,...,...,...,...,...,...
140782,"999 Wilson St, Atlanta, GA 30301",1,99.99,1,Atlanta (GA),False,99.99
140783,"999 Wilson St, Los Angeles, CA 90001",1,999.99,1,Los Angeles (CA),False,999.99
140784,"999 Wilson St, New York City, NY 10001",1,150.00,1,New York City (NY),False,150.00
140785,"999 Wilson St, Portland, OR 97035",1,2.99,1,Portland (OR),False,2.99


In [48]:
is_repeat = household_df.groupby('is_repeat').agg(households=('purchase_address', 'count'),
                                      avg_orders=('total_orders', 'mean'),
                                      avg_revenue=('total_revenue', 'mean'),
                                      median_revenue=('total_revenue', 'median'),
                                      avg_order_value=('avg_order_value', 'mean'))
is_repeat

,households,avg_orders,avg_revenue,median_revenue,avg_order_value
is_repeat,,,,,
False,109998,1.000000,193.992725,14.95,193.992725
True,30789,2.222839,427.205306,179.90,191.642594


## Statistical Validation: Repeat vs Non-Repeat Household Revenue

To check whether repeat households generate meaningfully different revenue from one-time households, I compared total household revenue between two groups:

- one-time households
- repeat households

Null hypothesis: there is no difference in average total revenue between the two groups.

Alternative hypothesis: repeat households generate different average total revenue.

In [30]:
repeat_revenue = household_df.loc[household_df['is_repeat'] == True,'total_revenue']

non_repeat_revenue = household_df.loc[ household_df['is_repeat'] == False,'total_revenue']

In [47]:
is_repeat.to_csv('repeat.csv', index=False)

In [31]:
from scipy.stats import ttest_ind

t_stat, p_value = ttest_ind(repeat_revenue,non_repeat_revenue,equal_var=False)

print("t-statistic:", t_stat)
print("p-value:", p_value)

t-statistic: 74.32868166833558
p-value: 0.0


In [32]:
if p_value < 0.05:
    print("Result: statistically significant difference")
else:
    print("Result: no statistically significant difference")

Result: statistically significant difference


In [33]:
mean_difference = repeat_revenue.mean() - non_repeat_revenue.mean()

percent_lift = (mean_difference / non_repeat_revenue.mean()) * 100

print("Mean difference:", mean_difference)
print("Percent lift:", percent_lift)

Mean difference: 233.2125803778399
Percent lift: 120.21717818599402


In [34]:
household_df['repeat_segment'] = pd.cut(household_df['total_orders'], bins=[0,1,2,999], 
                                         labels=['1 order', '2 orders', '3+ orders' ], include_lowest=True)
household_df[['purchase_address', 'total_orders', 'repeat_segment']].head()

,purchase_address,total_orders,repeat_segment
0,"1 11th St, Atlanta, GA 30301",1,1 order
1,"1 11th St, Los Angeles, CA 90001",1,1 order
2,"1 11th St, San Francisco, CA 94016",1,1 order
3,"1 12th St, Los Angeles, CA 90001",1,1 order
4,"1 12th St, New York City, NY 10001",1,1 order


In [35]:
repeat_segment_summary= household_df.groupby('repeat_segment', observed= False).agg(household=('purchase_address', 'count'),
                                                                                      avg_orders=('total_orders', 'mean'),
                                                                                      avg_revenue=('total_revenue', 'mean'),
                                                                                      median_revenue=('total_revenue', 'median'),
                                                                                      avg_order_value=('avg_order_value', 'mean'),
                                                                                      total_revenue=('total_revenue', 'sum')).reset_index()
repeat_segment_summary                                                  

,repeat_segment,household,avg_orders,avg_revenue,median_revenue,avg_order_value,total_revenue
0,1 order,109998,1.000000,193.992725,14.95,193.992725,21338811.81
1,2 orders,25043,2.000000,380.778688,161.99,190.389344,9535840.68
2,3+ orders,5746,3.194048,629.548117,413.93,197.104678,3617383.48


In [36]:
repeat_segment_summary['revenue_share_pct']=(repeat_segment_summary['total_revenue'] / repeat_segment_summary['total_revenue'].sum()) * 100

repeat_segment_summary['household_share_pct'] =(repeat_segment_summary['household'] / repeat_segment_summary['household'].sum()) *100

repeat_segment_summary

,repeat_segment,household,avg_orders,avg_revenue,median_revenue,avg_order_value,total_revenue,revenue_share_pct,household_share_pct
0,1 order,109998,1.000000,193.992725,14.95,193.992725,21338811.81,61.86591,78.130793
1,2 orders,25043,2.000000,380.778688,161.99,190.389344,9535840.68,27.64650,17.787864
2,3+ orders,5746,3.194048,629.548117,413.93,197.104678,3617383.48,10.48759,4.081343


In [59]:
repeat_city = household_df.groupby('city').agg(households=('purchase_address','count'),
                                                     repeat_households=('is_repeat','sum'),
                                                     repeat_rate=('is_repeat','mean'),
                                                     total_revenue=('total_revenue','sum')
                                                    )

repeat_city['repeat_rate'] = repeat_city['repeat_rate'] * 100

repeat_city = repeat_city.sort_values('repeat_rate', ascending=False).reset_index()
repeat_city

,city,households,repeat_households,repeat_rate,total_revenue
0,San Francisco (CA),28324,10847,38.296145,8262203.91
1,Los Angeles (CA),21450,5821,27.137529,5452570.80
2,New York City (NY),18807,4247,22.582017,4664317.43
3,Boston (MA),15706,2949,18.776264,3661642.01
4,Seattle (WA),12212,1734,14.199148,2747755.48
5,Atlanta (GA),12334,1728,14.010054,2795498.58
6,Dallas (TX),12321,1725,14.000487,2767975.40
7,Austin (TX),8609,841,9.768847,1819581.75
8,Portland (OR),8723,836,9.583859,1870732.34
9,Portland (ME),2301,61,2.651021,449758.27


In [56]:
repeat_city.to_csv('repeat_city.csv', index=False)

In [38]:
high_repeat_city = household_df.groupby('city').agg(households=('purchase_address', 'count'),
                                                          high_repeat_households_three_orders_plus=('total_orders', lambda x: (x >= 3).sum()),
                                                          high_repeat_rate_three_orders_plus=('total_orders', lambda x: (x >= 3).mean() * 100),
                                                          total_revenue=('total_revenue', 'sum')
                                                         ).sort_values('high_repeat_rate_three_orders_plus', ascending=False).reset_index()

high_repeat_city

,city,households,high_repeat_households_three_orders_plus,high_repeat_rate_three_orders_plus,total_revenue
0,San Francisco (CA),28324,2975,10.503460,8262203.91
1,Los Angeles (CA),21450,1061,4.946387,5452570.80
2,New York City (NY),18807,688,3.658212,4664317.43
3,Boston (MA),15706,396,2.521329,3661642.01
4,Dallas (TX),12321,182,1.477153,2767975.40
5,Atlanta (GA),12334,172,1.394519,2795498.58
6,Seattle (WA),12212,159,1.301998,2747755.48
7,Austin (TX),8609,56,0.650482,1819581.75
8,Portland (OR),8723,56,0.641981,1870732.34
9,Portland (ME),2301,1,0.043459,449758.27


## Findings

- the housholds that repeated purchases gererates significant higher total revenue around 120%, than the one time purchase households, and the differece is both statistically significant and commercially meaningful. 

- one order households represented 78.1% of all households but generated only 61.9% of total revenue.

- households with 3+ orders represented only 4.1% of households while contributing 10.5% of total revenue.

- average and median household revenue increased across repeat segments, while average order value stayed relatively stable. which suggests that higher cumulative revenue is mainly linked to repeated purchasing frequency rather than bigger individual orders.

- california cities demonstrated the strongest repeat purchasing behavior placing more than one order and high repeat purshasing behavior placing 3 or more order with 65% and 15.5% respectively.

- The business may benefit from focusing more on retaining repeat households, because a smaller group of repeat purchasers contributes disproportionately to revenue.

## Limitation

- purchase address acts only as a household proxy and cannot confirm unique customer identity which may affect the understanding of customer behavior.
- no compaign data, cost per acquisition data or retention cost per client/ household to determine how much from the budget would both acquiring new customer/houshold and retaining them should have. 
